packages/opentelemetry-exporter-datadog/src/datadogSpanProcessor.ts

```diff
@@ -0,0 +1,203 @@
+    this._bufferTimeout =
+      config && typeof config.bufferTimeout === 'number'
+        ? config.bufferTimeout
+        : DEFAULT_BUFFER_TIMEOUT_MS;
+
+    this._maxQueueSize =
+      config && typeof config.maxQueueSize === 'number'
+        ? config.maxQueueSize
+        : DEFAULT_MAX_QUEUE_SIZE;
+    this._maxTraceSize =
+      config && typeof config.maxTraceSize === 'number'
+        ? config.maxTraceSize
+        : DEFAULT_MAX_TRACE_SIZE;
+  }
+
+  forceFlush(): void {
+    if (this._isShutdown) {
+      return;
+    }
+    this._flush();
+  }
+
+  shutdown(): void {
+    if (this._isShutdown) {
+      return;
+    }
+    this.forceFlush();
+    this._isShutdown = true;
+    this._exporter.shutdown();
+  }
+
+  // does nothing.
+  onStart(span: ReadableSpan): void {
+    if (this._isShutdown) {
+      return;
+    }
+
+    if (this._allSpansCount(this._traces_spans_started) >= this._maxQueueSize) {
+      this.logger.debug('Max Queue Size reached, dropping span');
+      return;
+    }
+
+    const traceId = span.spanContext['traceId'];
+
+    if (!this._traces.has(traceId)) {
+      this._traces.set(traceId, []);
+      // this._traces.get(traceId).set(SPANS_KEY, [])
+      this._traces_spans_started.set(traceId, 0);
+      this._traces_spans_finished.set(traceId, 0);
+    }
+
+    const trace = this._traces.get(traceId);
+    const traceSpansStarted = this._traces_spans_started.get(traceId);
+
+    if (traceSpansStarted >= this._maxTraceSize) {
+      this.logger.debug('Max Trace Size reached, dropping span');
+      return;
+    }
+
+    trace.push(span);
+    this._traces_spans_started.set(traceId, traceSpansStarted + 1);
+  }
+
+  onEnd(span: ReadableSpan): void {
+    if (this._isShutdown) {
+      return;
+    }
+
+    const traceId = span.spanContext['traceId'];
+
+    if (!this._traces.has(traceId)) {
+      return;
+    }
+
+    const traceSpansFinished = this._traces_spans_finished.get(traceId);
+
+    this._traces_spans_finished.set(traceId, traceSpansFinished + 1);
+
+    if (this._isExportable(traceId)) {
+      this._check_traces_queue.add(traceId);
+      this._maybeStartTimer();
+    }
+  }
+
+  getTraceContext(span: ReadableSpan): string | undefined[] {
+    return [
+      id(span.spanContext['traceId']),
+      id(span.spanContext['spanId']),
+      span.parentSpanId ? id(span.parentSpanId) : undefined,
+    ];
+  }
+
+  /** Send the span data list to exporter */
+  private _flush() {
+    this._clearTimer();
+    if (this._check_traces_queue.size === 0) return;
+
+    this._check_traces_queue.forEach(traceId => {
+      // check again in case spans have been added
+      if (this._isExportable(traceId)) {
+        const spans = this._traces.get(traceId);
+        this._traces.delete(traceId);
+        this._traces_spans_started.delete(traceId);
+        this._traces_spans_finished.delete(traceId);
+        this._check_traces_queue.delete(traceId);
+        this._exporter.export(spans, () => {});
```

Comment:
you might want to log an error here

Reply:
I agree this will make an export fail silently which is no good

Reply:
yup makes sense, i've added an error log here. (are there preferences on log level? error feels appropriate here since most debug logs wont get registered, but not sure if there's otel conventions are noiseyness)

Reply:
The exporter for the otel collector logs an error for reference, so i think thats the way to go

Reply:
ty, will keep as error log.